### PR TITLE
[ENH] fit second level model on surface

### LINKED ISSUE #4617: support 2nd level GLM analysis of SurfaceImage objects
Realizing that we do not support 2nd level analysis of SurfaceImage objects?

@bthirion This is something we want to have, right?

_Originally posted by @Remi-Gau in https://github.com/nilearn/nilearn/issues/4603#issuecomment-2397105199_

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes #4617

Changes proposed in this pull request:

- move surface testing fixture to proprer functions to the `nilearn.surface._testing`
- update examples/04_glm_first_level/plot_localizer_surface_analysis.py to use SecondLevelModel
- update SecondLevelModel to handle surface images

### CODE REVIEW

File: nilearn/glm/tests/test_first_level.py [hunk 1]
```diff
@@ -2228,6 +2213,28 @@ def test_flm_compute_contrast_with_surface_data(_make_surface_glm_data):
     assert isinstance(result, SurfaceImage)
 
 
+def test_flm_get_voxelwise_model_attribute_with_surface_data(
+    surface_glm_data,
+):
+    """Smoke test 'voxel wise' attribute with surface data.
+
+    TODO: rename the private function _get_voxelwise_model_attribute
+    to work for both voxel and vertex
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:2222]: TODO: private function should be renamed

**Comment by Remi-Gau** [comment]
- blocked by https://github.com/nilearn/nilearn/issues/4592

**Comment by Remi-Gau** [comment]
OK seems the SurfaceMasker is doing something wrong when you pass a list of images to transform

- so something was missed in https://github.com/nilearn/nilearn/pull/4719

**Review by bthirion — COMMENTED** [review]
The PR LGTM.

File: nilearn/conftest.py [hunk 2]
```diff
@@ -508,29 +509,6 @@ def f(img):
...
-def assert_surf_mesh_equal():
-    """Check that 2 meshes are equal."""
-
-    def f(mesh_1, mesh_2):
-        assert np.array_equal(mesh_1.coordinates, mesh_2.coordinates)
-        assert np.array_equal(mesh_1.faces, mesh_2.faces)
-
-    return f
-
-
```
Comment (Remi-Gau) [inline nilearn/conftest.py:533]: move surface testing fixture to proprer functions to the `nilearn.surface.testing`

**Comment by Remi-Gau** [comment]
> So the issue wasn't with the SurfaceMasker?

yes and know: I think we have an issue with PolyData that is completely fine have a 1 dim np.array for each parts.

I will open a separate PR to enforce 2 dimensions for the parts in polydata: that should prevent some headaches down the line.

File: nilearn/glm/tests/test_second_level.py [hunk 3]
```diff
@@ -198,6 +198,18 @@ def test_second_level_input_as_3d_images(
...
+    design_matrix = pd.DataFrame(
+        [1] * img_4d_mni.shape[3], columns=["intercept"]
+    )
+    second_level_model = SecondLevelModel()
+    second_level_model = second_level_model.fit(
+        second_level_input=img_4d_mni,
+        design_matrix=design_matrix,
+    )
+
+
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_second_level.py:212]:
```suggestion
```

File: examples/07_advanced/plot_surface_bids_analysis.py [hunk 4]
```diff
@@ -136,9 +157,35 @@
...
+        title=f"(language-string), {hemi} hemisphere ; scipy",
+        colorbar=True,
+        cmap="bwr",
+        threshold=1.96,
+        bg_map=fsaverage_data,
+    )
+
+
+# %%
+# Use SecondLevel
```
Comment (Remi-Gau) [inline examples/07_advanced/plot_surface_bids_analysis.py:169]: @bthirion  
Any idea why using scipy and Second level model would give such different results?
Am I missing something obvvious?

## OLD: pure scipy

Figure_1_scipy

## NEW: Nilearn Second level model

Figure_3_nl

Once this discrepancy I would prefer to only keep the nilearn second level model appraoch in this example.

**Comment by Remi-Gau** [comment]
> the two seem to coincide.

they do !!

File: examples/07_advanced/plot_surface_bids_analysis.py [hunk 5]
```diff
@@ -47,28 +47,49 @@
...
 # with their subject_id filled from the :term:`BIDS` dataset.
-# Moreover, we obtain for each model
-# a dictionary with run_imgs, events and confounder regressors
-# since in this case a confounds.tsv file is available
-# in the :term:`BIDS` dataset.
+# Along, we also obtain:
+#
+#   - a list with the Nifti image associated with each run
+#
+#   - a list of events read from events.tsv in the the :term:`BIDS` dataset
```
Comment (man-shu) [inline examples/07_advanced/plot_surface_bids_analysis.py:54]:
```suggestion
#   - a list of events read from events.tsv in the :term:`BIDS` dataset
```

File: examples/07_advanced/plot_surface_bids_analysis.py [hunk 6]
```diff
@@ -47,28 +47,49 @@
...
 # The projection function simply takes the :term:`fMRI` data and the mesh.
 # Note that those correspond spatially, as they are both in :term:`MNI` space.
+#
+# .. warning::
+#
+#    Note that here we pass ALL the confounds when we fit the model.
+#    In this case we can do this because our regressors only include
+#    the motion realignment parameters.
+#    For most preprocessed BIDS dataset,
+#    you would have to carefully choose which confounds to include.
```
Comment (man-shu) [inline examples/07_advanced/plot_surface_bids_analysis.py:91]: How about also mentioning the [`load_confounds`](https://nilearn.github.io/stable/modules/generated/nilearn.interfaces.fmriprep.load_confounds.html#nilearn-interfaces-fmriprep-load-confounds) function?

File: examples/07_advanced/plot_surface_bids_analysis.py [hunk 7]
```diff
@@ -90,55 +112,55 @@
...
 # %%
 # Group study
 # -----------
 #
-# Individual activation maps have been accumulated
-# in the ``z_score_left`` and ``z_scores_right`` lists respectively.
+# Individual activation maps have been accumulated in the ``z_score``.
 # We can now use them in a group study (one-sample study).
+# We can now pass this to a second level model for a one-sample t-test
+# to fit and compute contrast on.
```
Comment (man-shu) [inline examples/07_advanced/plot_surface_bids_analysis.py:133]:
```suggestion
# We can now use them in a group study (one-sample study) by passing ``z_score`` to a second level model for a one-sample t-test to fit and compute contrast on.
```

How about a little bit of rephrasing?

File: nilearn/glm/tests/test_first_level.py [hunk 8]
```diff
@@ -2235,24 +2223,47 @@ def test_flm_with_surface_masker_with_mask(_make_surface_glm_data, surf_mask):
...
+    """
+    img, _ = surface_glm_data(5)
+    masker = SurfaceMasker().fit(img)
+    model = FirstLevelModel(mask_img=masker, t_r=2.0, minimize_memory=False)
+    events = basic_paradigm()
+    model.fit([img, img], events=[events, events])
+
+    model.residuals  # noqa: B018
+    model.predicted  # noqa: B018
+    model.r_square  # noqa: B018
```
Comment (man-shu) [inline nilearn/glm/tests/test_first_level.py:2263]: Should we be checking something here with `assert` statements?

**Review by man-shu — COMMENTED** [review]
Not sure if you wanted a review yet, but thought would be good to catch up with changes so far.

File: examples/07_advanced/plot_surface_bids_analysis.py (hunk 6 above)
Comment (Remi-Gau) [reply examples/07_advanced/plot_surface_bids_analysis.py:91]: good point

Comment (Remi-Gau) [reply examples/07_advanced/plot_surface_bids_analysis.py:91]: added something

File: examples/07_advanced/plot_surface_bids_analysis.py (hunk 7 above)
Comment (Remi-Gau) [reply examples/07_advanced/plot_surface_bids_analysis.py]: rephrased more or less the whole passage: LMKWYT

**Comment by Remi-Gau** [comment]
> Not sure if you wanted a review yet, but thought would be good to catch up with changes so far.

that was perfect thanks !

File: nilearn/glm/second_level/second_level.py [hunk 9]
```diff
@@ -76,21 +87,16 @@ def _check_input_type_when_list(second_level_input):
...
-        "second_level_input must be "
-        "either a pandas DataFrame, "
-        "a Niimg-like object, "
-        "a list of Niimg-like object or "
-        "a list of FirstLevelModel objects. "
-        f"Got {_return_type(second_level_input)} instead"
-    )
+    if all(isinstance(x, SurfaceImage) for x in second_level_input):
+        return "surf_img_object"
+    raise TypeError(_input_type_error_message(second_level_input))
```
Comment (man-shu) [inline nilearn/glm/second_level/second_level.py:99]: There are some more tests required...

**Review by bthirion — COMMENTED** [review]
Almost there. LMK if you need some help on it.

File: nilearn/glm/tests/test_first_level.py (hunk 8 above)
Comment (Remi-Gau) [reply nilearn/glm/tests/test_first_level.py]: fixed

File: nilearn/glm/second_level/second_level.py (hunk 9 above)
Comment (Remi-Gau) [reply nilearn/glm/second_level/second_level.py:106]: fixed

File: nilearn/maskers/_utils.py [hunk 10]
```diff
@@ -63,7 +63,8 @@ def compute_mean_surface_image(img):
     -------
     SurfaceImage
     """
-    if img.shape[0] < 2:
+    n_timpoints = img.shape[0]
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply nilearn/maskers/_utils.py]: fixed

File: examples/07_advanced/plot_surface_bids_analysis.py [hunk 11]
```diff
@@ -42,33 +42,62 @@
...
 models, run_imgs, events, confounds = first_level_from_bids(
-    data.data_dir,
-    task_label,
+    dataset_path=data.data_dir,
+    task_label="languagelocalizer",
     img_filters=[("desc", "preproc")],
+    sub_labels=["01", "02", "03", "04", "05"],  # comment to run all subjects
     hrf_model="glover + derivative",
-    n_jobs=2,
+    n_jobs=-1,
```
[the remark this answers is not in the document]
Comment (Remi-Gau) [reply examples/07_advanced/plot_surface_bids_analysis.py]: fixed

File: nilearn/maskers/_utils.py [hunk 12]
```diff
@@ -125,3 +128,34 @@ def concatenate_surface_images(imgs):
     output = SurfaceImage(mesh=imgs[0].mesh, data=output_data)
 
     return output
+
+
+def deconcatenate_surface_images(img):
```
Comment (Remi-Gau) [inline nilearn/maskers/_utils.py:133]: the surface utility functions from this module should probably be moved somewhere else (nilearn._utils.surface ?)

**Comment by Remi-Gau** [comment]
- The doc build error is fixed as part of https://github.com/nilearn/nilearn/pull/4790

File: nilearn/glm/tests/test_second_level.py [hunk 13]
```diff
@@ -1384,3 +1394,148 @@ def test_second_lvl_dataframe_computation(tmp_path, shape_3d_default):
...
+    second_level_input = [surf_img() for _ in range(n_subjects)]
+
+    design_matrix = pd.DataFrame(
+        [1] * len(second_level_input), columns=["intercept"]
+    )
+
+    model = SecondLevelModel(smoothing_fwhm=8.0)
+    # TODO
+    # should throw warning that smoothing is not implemented
+    model = model.fit(second_level_input, design_matrix=design_matrix)
```
Comment (man-shu) [inline nilearn/glm/tests/test_second_level.py:1483]: Maybe a check for the warning is needed here?

**Review by man-shu — APPROVED** [review]
Otherwise LGTM!

File: nilearn/glm/tests/test_second_level.py (hunk 13 above)
Comment (Remi-Gau) [reply nilearn/glm/tests/test_second_level.py:1483]: good catch!!
thanks

Comment (Remi-Gau) [reply nilearn/glm/tests/test_second_level.py]: fixed

File: nilearn/glm/tests/test_first_level.py (hunk 1 above)
Comment (bthirion) [reply nilearn/glm/tests/test_first_level.py:2252]: you leave it for a future PR ?

File: nilearn/maskers/tests/test_utils.py [hunk 14]
```diff
@@ -48,3 +55,36 @@ def test_concatenate_surface_images(surf_img):
...
+    for i in range(input.shape[1]):
+        assert_polymesh_equal(output[i].mesh, input.mesh)
+        assert_array_equal(
+            np.squeeze(output[i].data.parts["left"]),
+            input.data.parts["left"][..., i],
+        )
+
+
+def test_deconcatenate_surface_images_2d(surf_img):
+    """Retunn as is if surface image is 2D."""
```
Comment (bthirion) [inline nilearn/maskers/tests/test_utils.py:76]: typo

File: nilearn/glm/second_level/second_level.py [hunk 15]
```diff
@@ -202,10 +218,31 @@ def _check_input_as_nifti_images(second_level_input, none_design_matrix):
...
+            " require a design matrix to be provided."
         )
 
 
+def _check_input_as_surface_images(second_level_input, none_design_matrix):
+    if (
+        isinstance(second_level_input, SurfaceImage)
+        and second_level_input.shape[0] == 1
+    ):
+        raise TypeError(
```
Comment (bthirion) [inline nilearn/glm/second_level/second_level.py:230]: line not covered ?

**Review by bthirion — COMMENTED** [review]
LGTM, only 2 comments remain.

File: nilearn/glm/tests/test_first_level.py (hunk 1 above)
Comment (Remi-Gau) [reply nilearn/glm/tests/test_first_level.py:2252]: yes: this PR is big enough

File: nilearn/maskers/tests/test_utils.py (hunk 14 above)
Comment (Remi-Gau) [inline nilearn/maskers/tests/test_utils.py:76]:
```suggestion
    """Return as is if surface image is 2D."""
```

**Comment by Remi-Gau** [comment]
- blocked by #4802

**Review by bthirion — APPROVED** [review]
LGTM, thx.
